src/libsyntax/ast.rs

```diff
@@ -1746,13 +1730,21 @@ impl StructFieldKind {
     }
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum VariantKind {
+    Dict,
```

Comment:
The problem is that most of code is not interested in named-ness of fields, so we'd have to return to all the code duplication
```
match kind {
    Struct(named_fields, _) => {
        // loop through named fields and do thing
    }
    Tuple(unnamed_fields) => {
        // loop through unnamed fields and do the same thing
    }
    _ => {} // do not loop
}
```
which I so happily eliminated in the first two commits. Oh, and `Unit` needs ctor_id too for the constant it introduces.

```
enum DataKind {
    Unit(NodeId), // NodeId == ctor id
    Struct(Vec<StructField>),
    Tuple(Vec<StructField>, NodeId), // NodeId == ctor id
}
```
could work out if we could write a getter for fields: 
`fn fields(data_kind: &DataKind) -> &[StructField] { ... }`,
but `Unit` can't provide a reference to `[StructField]` unfortunately.
And if we add fields to `Unit` we get
```
enum DataKind {
    Unit(Vec<StructField>, NodeId), // empty vector, NodeId == ctor id
    Struct(Vec<StructField>),
    Tuple(Vec<StructField>, NodeId), // NodeId == ctor id
}
```
and that is almost what we have in this PR.